Hooks.ts

```diff
@@ -1,28 +1,57 @@
-import * as cloudflare from "@cloudflare/workers-types"
 import { http } from "cloudly-http"
 import { storage } from "cloudly-storage"
 import { Queue } from "./Queue"
+import { Types } from "./Types"
 
-export class Hooks {
-	private constructor(readonly queue: Queue) {}
-	async trigger(target: string, value: any): Promise<void> {
-		let request: http.Request
-		if (!http.Request.is(value)) {
-			request = http.Request.create({
-				header: {
-					contentType: "application/json",
-				},
-				method: "POST",
-				body: value,
-				url: target,
-			})
-		} else {
-			request = value
+export class Hooks<T extends string> {
+	#options?: Hooks.Options
+	private constructor(
+		readonly queue: Queue,
+		readonly registrations?: storage.KeyValueStore<{ url: string; header?: http.Request.Header }>
+	) {}
+	async register(listener: Hooks.Registration<T>): Promise<boolean> {
+		return (
+			this.registrations
+				?.set(`${listener.hook}|${new URL(listener.url).hostname}`, {
+					url: listener.url,
+					header: listener.header,
+				})
+				.then(
+					() => true,
+					() => false
+				) ?? false
+		)
+	}
+	configure(maxRetries: number, timeFactor: number) {
+		this.#options = { maxRetries, timeFactor }
+	}
+	async trigger(hook: T, body: any, destinations?: string[]): Promise<void> {
+		const options = this.#options
+		const events: Hooks.Event<T>[] = []
+		if (destinations)
+			events.push(...destinations.map(d => ({ hook, url: d, body, options })))
+		else {
+			events.push(
+				...((await this.registrations?.list())?.flatMap(r =>
+					r.key.split("|")[0] == hook && r.value ? [{ ...r.value, body, hook, options }] : []
+				) ?? [])
+			)
 		}
-		this.queue.enqueue(request)
+		await Promise.all(events.map(async event => await this.queue.enqueue(event)))
```

Comment:
This will make 1 DO call per listener we have, if we make it so each DO handle an array of listeners we can reduce the amount of calls our worker invocation needs to make  and move them to the DO. But the amount of listeners is probably low. I can't decide on this...